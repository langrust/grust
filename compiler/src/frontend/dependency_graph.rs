mod component;
mod expression;
mod file;
mod statement;
mod stream_expression;
