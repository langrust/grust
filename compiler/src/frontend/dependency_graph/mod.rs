use petgraph::graphmap::DiGraphMap;

use crate::common::label::Label;

mod contract;
mod expression;
mod file;
mod node;
mod stream_expression;
mod statement;

/// Add an egde to the graph.
///
/// If edge already exits then keep the edge with the lowest weight.
pub fn add_edge(
    graph: &mut DiGraphMap<usize, Label>,
    signal_id: usize,
    dependency_id: usize,
    label: Label,
) {
    let prev_label = graph.add_edge(signal_id, dependency_id, label.clone());
    match (prev_label, label) {
        (Some(Label::Weight(prev_weight)), Label::Weight(new_weight))
            if prev_weight < new_weight =>
        {
            graph.add_edge(signal_id, dependency_id, Label::Weight(prev_weight));
        }
        _ => (),
    }
}
