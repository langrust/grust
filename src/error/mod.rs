use codespan_reporting::diagnostic::{Diagnostic, Label};

use crate::ast::{
    location::Location,
    type_system::Type,
};

/// Compilation errors enumeration.
/// 
/// [Error] enumeration is used during the compilation to alert
/// the programmer of some errors in its LanGRust program.
/// 
/// # Example
/// ```rust
/// use grustine::error::Error;
/// use grustine::ast::location::Location;
/// 
/// let mut errors = vec![];
/// 
/// let name = String::from("unknown");
/// let location = Location::default();
/// 
/// let error = Error::UnknownElement { name, location };
/// errors.push(error);
/// ```
/// 
#[derive(Debug, PartialEq, Eq, Clone)]
pub enum Error {
    /// encountering an unknown element
    UnknownElement {
        /// the unknow identifier
        name: String,
        /// the error location
        location: Location,
    },
    /// incompatible application
    IncompatibleInputType {
        /// given type as input
        given_type: Type,
        /// expected type as input
        expected_type: Type,
        /// the error location
        location: Location,
    },
}
impl Error {
    /// Transform the error into a diagnostic.
    /// 
    /// This makes it possible to use [codespan_reporting] to print
    /// pretty errors.
    /// 
    /// # Example
    /// ```rust
    /// use grustine::error::Error;
    /// use grustine::ast::location::Location;
    /// use codespan_reporting::{
    ///     files::SimpleFiles,
    ///     term::termcolor::{StandardStream, ColorChoice},
    ///     term,
    /// };
    /// 
    /// let mut errors: Vec<Error> = vec![];
    /// let mut files = SimpleFiles::new();
    /// 
    /// let file_id = files.add("file_test.gr", "a code without x...");
    /// let name = String::from("x");
    /// let location = Location {
    ///     file_id,
    ///     range: 0..0,
    /// };
    /// 
    /// let error = Error::UnknownElement { name, location };
    /// errors.push(error);
    /// 
    /// let writer = StandardStream::stderr(ColorChoice::Always);
    /// let config = term::Config::default();
    /// for error in &errors {
    ///     let writer = &mut writer.lock();
    ///     let _ = term::emit(writer, &config, &files, &error.to_diagnostic());
    /// }
    /// ```
    pub fn to_diagnostic(&self) -> Diagnostic<usize> {
        match self {
            Error::UnknownElement { name, location } => Diagnostic::error()
                .with_message("unknown element")
                .with_labels(vec![
                    Label::primary(location.file_id, location.range.clone())
                        .with_message("unknown")
                ])
                .with_notes(vec![
                    format!("element '{name}' is not defined")
                ]
            ),
            Error::IncompatibleInputType { given_type, expected_type, location } => Diagnostic::error()
                .with_message("incompatible application")
                .with_labels(vec![
                    Label::primary(location.file_id, location.range.clone())
                        .with_message("wrong input type")
                ])
                .with_notes(vec![
                    format!("expected '{expected_type}' but '{given_type}' was given")
                ]
            ),
        }
    }
}
