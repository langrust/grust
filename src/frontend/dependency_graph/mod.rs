use petgraph::graphmap::DiGraphMap;

use crate::common::label::Label;

mod expression;
mod file;
mod node;
mod stream_expression;
mod term;

pub fn add_edge(
    graph: &mut DiGraphMap<usize, Label>,
    signal_id: usize,
    dependency_id: usize,
    label: Label,
) {
    let prev_label = graph.add_edge(signal_id, dependency_id, label.clone());
    match (prev_label, label) {
        (Some(Label::Weight(prev_weight)), Label::Weight(new_weight))
            if prev_weight < new_weight =>
        {
            graph.add_edge(signal_id, dependency_id, Label::Weight(prev_weight));
        }
        _ => (),
    }
}
