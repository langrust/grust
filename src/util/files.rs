use codespan_reporting::files;
use std::ops::Range;

/// A file that is backed by an `Arc<String>`.
#[derive(Debug, Clone)]
struct File {
    /// The name of the file.
    name: String,
    /// The source code of the file.
    source: String,
    /// The starting byte indices in the source code.
    line_starts: Vec<usize>,
}

impl File {
    fn line_start(&self, line_index: usize) -> Result<usize, files::Error> {
        use std::cmp::Ordering;

        match line_index.cmp(&self.line_starts.len()) {
            Ordering::Less => Ok(*self
                .line_starts
                .get(line_index)
                .expect("failed despite previous check")),
            Ordering::Equal => Ok(self.source.len()),
            Ordering::Greater => Err(files::Error::LineTooLarge {
                given: line_index,
                max: self.line_starts.len() - 1,
            }),
        }
    }
}

/// An opaque file identifier.
#[derive(Copy, Clone, PartialEq, Eq, Debug)]
pub struct FileId(i32);
impl FileId {
    /// Construct and return the default file identifier.
    pub fn default() -> Self {
        return FileId(-1);
    }
}

/// The files handler.
#[derive(Debug, Clone)]
pub struct Files {
    files: Vec<File>,
}

impl Files {
    /// Create a new files database.
    pub fn new() -> Files {
        Files { files: Vec::new() }
    }

    /// Add a file to the database, returning the handle that can be used to
    /// refer to it again.
    pub fn add(
        &mut self,
        name: impl Into<String>,
        source: impl Into<String>,
    ) -> Option<FileId> {
        let file_id = FileId(i32::try_from(self.files.len()).ok()?);
        let name = name.into();
        let source = source.into();
        let line_starts = files::line_starts(&source).collect();

        self.files.push(File {
            name,
            line_starts,
            source,
        });

        Some(file_id)
    }

    /// Get the file corresponding to the given id.
    fn get(&self, file_id: FileId) -> Result<&File, files::Error> {
        self.files
            .get(file_id.0 as usize)
            .ok_or(files::Error::FileMissing)
    }
}

impl<'files> files::Files<'files> for Files {
    type FileId = FileId;
    type Name = &'files str;
    type Source = &'files str;

    fn name(&self, file_id: FileId) -> Result<&str, files::Error> {
        Ok(self.get(file_id)?.name.as_ref())
    }

    fn source(&self, file_id: FileId) -> Result<&str, files::Error> {
        Ok(&self.get(file_id)?.source)
    }

    fn line_index(
        &self,
        file_id: FileId,
        byte_index: usize
    ) -> Result<usize, files::Error> {
        self.get(file_id)?
            .line_starts
            .binary_search(&byte_index)
            .or_else(|next_line| Ok(next_line - 1))
    }

    fn line_range(
        &self,
        file_id: FileId,
        line_index: usize,
    ) -> Result<Range<usize>, files::Error> {
        let file = self.get(file_id)?;
        let line_start = file.line_start(line_index)?;
        let next_line_start = file.line_start(line_index + 1)?;

        Ok(line_start..next_line_start)
    }
}