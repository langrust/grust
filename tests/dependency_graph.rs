use codespan_reporting::files::SimpleFiles;

use grustine::dependency_graph;
use grustine::error::display;

#[test]
fn dependency_graph_of_counter() {
    let mut files = SimpleFiles::new();

    let counter_id = files.add(
        "counter.gr",
        std::fs::read_to_string("tests/fixture/dependency_graph/success/counter.gr")
            .expect("unkown file"),
    );

    match dependency_graph(counter_id, &mut files) {
        Ok(file) => insta::assert_yaml_snapshot!(file),
        Err(errors) => display(&errors, &files),
    }
}

#[test]
fn dependency_graph_of_blinking() {
    let mut files = SimpleFiles::new();

    let blinking_id = files.add(
        "blinking.gr",
        std::fs::read_to_string("tests/fixture/dependency_graph/success/blinking.gr")
            .expect("unkown file"),
    );

    match dependency_graph(blinking_id, &mut files) {
        Ok(file) => insta::assert_yaml_snapshot!(file),
        Err(errors) => display(&errors, &files),
    }
}

#[test]
fn dependency_graph_of_button_management() {
    let mut files = SimpleFiles::new();

    let button_management_id = files.add(
        "button_management.gr",
        std::fs::read_to_string("tests/fixture/dependency_graph/success/button_management.gr")
            .expect("unkown file"),
    );

    match dependency_graph(button_management_id, &mut files) {
        Ok(file) => insta::assert_yaml_snapshot!(file),
        Err(errors) => display(&errors, &files),
    }
}

#[test]
fn dependency_graph_of_button_management_condition_match() {
    let mut files = SimpleFiles::new();

    let button_management_condition_match_id = files.add(
        "button_management_condition_match.gr",
        std::fs::read_to_string(
            "tests/fixture/dependency_graph/success/button_management_condition_match.gr",
        )
        .expect("unkown file"),
    );

    match dependency_graph(button_management_condition_match_id, &mut files) {
        Ok(file) => insta::assert_yaml_snapshot!(file),
        Err(errors) => display(&errors, &files),
    }
}

#[test]
fn dependency_graph_of_button_management_using_function() {
    let mut files = SimpleFiles::new();

    let button_management_using_function_id = files.add(
        "button_management_using_function.gr",
        std::fs::read_to_string(
            "tests/fixture/dependency_graph/success/button_management_using_function.gr",
        )
        .expect("unkown file"),
    );

    match dependency_graph(button_management_using_function_id, &mut files) {
        Ok(file) => insta::assert_yaml_snapshot!(file),
        Err(errors) => display(&errors, &files),
    }
}

#[test]
fn dependency_graph_of_pid() {
    let mut files = SimpleFiles::new();

    let pid_id = files.add(
        "pid.gr",
        std::fs::read_to_string("tests/fixture/dependency_graph/success/pid.gr")
            .expect("unkown file"),
    );

    match dependency_graph(pid_id, &mut files) {
        Ok(file) => insta::assert_yaml_snapshot!(file),
        Err(errors) => display(&errors, &files),
    }
}
