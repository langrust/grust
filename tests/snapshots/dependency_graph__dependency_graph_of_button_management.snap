---
source: tests/dependency_graph.rs
expression: file
---
typedefs:
  - Enumeration:
      id: Button
      elements:
        - Released
        - Pressed
      location:
        file_id: 0
        range:
          start: 0
          end: 42
  - Enumeration:
      id: ResetState
      elements:
        - InProgress
        - Confirmed
      location:
        file_id: 0
        range:
          start: 44
          end: 94
functions: []
nodes:
  - id: counter
    is_component: false
    inputs:
      - - res
        - Boolean
      - - inc
        - Integer
    unscheduled_equations:
      o:
        scope: Output
        id: o
        signal_type: Integer
        expression:
          MapApplication:
            function_expression:
              Call:
                id: if_then_else
                typing:
                  Abstract:
                    - - Boolean
                      - Integer
                      - Integer
                    - Integer
                location:
                  file_id: 0
                  range:
                    start: 149
                    end: 183
            inputs:
              - SignalCall:
                  signal:
                    id: res
                    scope: Input
                  typing: Boolean
                  location:
                    file_id: 0
                    range:
                      start: 152
                      end: 155
                  dependencies:
                    - - res
                      - 0
              - Constant:
                  constant:
                    Integer: 0
                  typing: Integer
                  location:
                    file_id: 0
                    range:
                      start: 161
                      end: 162
                  dependencies: []
              - MapApplication:
                  function_expression:
                    Call:
                      id: " + "
                      typing:
                        Abstract:
                          - - Integer
                            - Integer
                          - Integer
                      location:
                        file_id: 0
                        range:
                          start: 178
                          end: 179
                  inputs:
                    - MapApplication:
                        function_expression:
                          Call:
                            id: (_)
                            typing:
                              Abstract:
                                - - Integer
                                - Integer
                            location:
                              file_id: 0
                              range:
                                start: 168
                                end: 177
                        inputs:
                          - FollowedBy:
                              constant:
                                Integer: 0
                              expression:
                                SignalCall:
                                  signal:
                                    id: o
                                    scope: Output
                                  typing: Integer
                                  location:
                                    file_id: 0
                                    range:
                                      start: 175
                                      end: 176
                                  dependencies:
                                    - - o
                                      - 0
                              typing: Integer
                              location:
                                file_id: 0
                                range:
                                  start: 169
                                  end: 176
                              dependencies:
                                - - o
                                  - 1
                        typing: Integer
                        location:
                          file_id: 0
                          range:
                            start: 168
                            end: 177
                        dependencies:
                          - - o
                            - 1
                    - SignalCall:
                        signal:
                          id: inc
                          scope: Input
                        typing: Integer
                        location:
                          file_id: 0
                          range:
                            start: 180
                            end: 183
                        dependencies:
                          - - inc
                            - 0
                  typing: Integer
                  location:
                    file_id: 0
                    range:
                      start: 168
                      end: 183
                  dependencies:
                    - - o
                      - 1
                    - - inc
                      - 0
            typing: Integer
            location:
              file_id: 0
              range:
                start: 149
                end: 183
            dependencies:
              - - res
                - 0
              - - o
                - 1
              - - inc
                - 0
        location:
          file_id: 0
          range:
            start: 136
            end: 184
    unitary_nodes: {}
    location:
      file_id: 0
      range:
        start: 96
        end: 186
    graph:
      vertices:
        inc:
          id: inc
          value: Black
          neighbors: []
        o:
          id: o
          value: Black
          neighbors:
            - id: res
              weight: 0
            - id: o
              weight: 1
            - id: inc
              weight: 0
        res:
          id: res
          value: Black
          neighbors: []
  - id: reset_button
    is_component: false
    inputs:
      - - button_state
        - Enumeration: Button
      - - period
        - Integer
      - - reset_limit_1
        - Integer
      - - reset_limit_2
        - Integer
    unscheduled_equations:
      counter:
        scope: Local
        id: counter
        signal_type: Integer
        expression:
          NodeApplication:
            node: counter
            inputs:
              - SignalCall:
                  signal:
                    id: res
                    scope: Local
                  typing: Boolean
                  location:
                    file_id: 0
                    range:
                      start: 787
                      end: 790
                  dependencies:
                    - - res
                      - 0
              - SignalCall:
                  signal:
                    id: period
                    scope: Input
                  typing: Integer
                  location:
                    file_id: 0
                    range:
                      start: 792
                      end: 798
                  dependencies:
                    - - period
                      - 0
            signal: o
            typing: Integer
            location:
              file_id: 0
              range:
                start: 779
                end: 801
            dependencies:
              - - res
                - 0
              - - res
                - 1
              - - period
                - 0
        location:
          file_id: 0
          range:
            start: 764
            end: 802
      res:
        scope: Local
        id: res
        signal_type: Boolean
        expression:
          MapApplication:
            function_expression:
              Call:
                id: " && "
                typing:
                  Abstract:
                    - - Boolean
                      - Boolean
                    - Boolean
                location:
                  file_id: 0
                  range:
                    start: 852
                    end: 854
            inputs:
              - MapApplication:
                  function_expression:
                    Call:
                      id: (_)
                      typing:
                        Abstract:
                          - - Boolean
                          - Boolean
                      location:
                        file_id: 0
                        range:
                          start: 819
                          end: 851
                  inputs:
                    - MapApplication:
                        function_expression:
                          Call:
                            id: " == "
                            typing:
                              Abstract:
                                - - Enumeration: Button
                                  - Enumeration: Button
                                - Boolean
                            location:
                              file_id: 0
                              range:
                                start: 833
                                end: 835
                        inputs:
                          - SignalCall:
                              signal:
                                id: button_state
                                scope: Input
                              typing:
                                Enumeration: Button
                              location:
                                file_id: 0
                                range:
                                  start: 820
                                  end: 832
                              dependencies:
                                - - button_state
                                  - 0
                          - Constant:
                              constant:
                                Enumeration:
                                  - Button
                                  - Pressed
                              typing:
                                Enumeration: Button
                              location:
                                file_id: 0
                                range:
                                  start: 836
                                  end: 850
                              dependencies: []
                        typing: Boolean
                        location:
                          file_id: 0
                          range:
                            start: 820
                            end: 850
                        dependencies:
                          - - button_state
                            - 0
                  typing: Boolean
                  location:
                    file_id: 0
                    range:
                      start: 819
                      end: 851
                  dependencies:
                    - - button_state
                      - 0
              - MapApplication:
                  function_expression:
                    Call:
                      id: (_)
                      typing:
                        Abstract:
                          - - Boolean
                          - Boolean
                      location:
                        file_id: 0
                        range:
                          start: 855
                          end: 897
                  inputs:
                    - FollowedBy:
                        constant:
                          Boolean: true
                        expression:
                          MapApplication:
                            function_expression:
                              Call:
                                id: " == "
                                typing:
                                  Abstract:
                                    - - Enumeration: Button
                                      - Enumeration: Button
                                    - Boolean
                                location:
                                  file_id: 0
                                  range:
                                    start: 878
                                    end: 880
                            inputs:
                              - SignalCall:
                                  signal:
                                    id: button_state
                                    scope: Input
                                  typing:
                                    Enumeration: Button
                                  location:
                                    file_id: 0
                                    range:
                                      start: 865
                                      end: 877
                                  dependencies:
                                    - - button_state
                                      - 0
                              - Constant:
                                  constant:
                                    Enumeration:
                                      - Button
                                      - Released
                                  typing:
                                    Enumeration: Button
                                  location:
                                    file_id: 0
                                    range:
                                      start: 881
                                      end: 896
                                  dependencies: []
                            typing: Boolean
                            location:
                              file_id: 0
                              range:
                                start: 865
                                end: 896
                            dependencies:
                              - - button_state
                                - 0
                        typing: Boolean
                        location:
                          file_id: 0
                          range:
                            start: 856
                            end: 896
                        dependencies:
                          - - button_state
                            - 1
                  typing: Boolean
                  location:
                    file_id: 0
                    range:
                      start: 855
                      end: 897
                  dependencies:
                    - - button_state
                      - 1
            typing: Boolean
            location:
              file_id: 0
              range:
                start: 819
                end: 897
            dependencies:
              - - button_state
                - 0
              - - button_state
                - 1
        location:
          file_id: 0
          range:
            start: 807
            end: 898
      user_reset_request:
        scope: Output
        id: user_reset_request
        signal_type: Boolean
        expression:
          Match:
            expression:
              SignalCall:
                signal:
                  id: button_state
                  scope: Input
                typing:
                  Enumeration: Button
                location:
                  file_id: 0
                  range:
                    start: 612
                    end: 624
                dependencies:
                  - - button_state
                    - 0
            arms:
              - - Constant:
                    constant:
                      Enumeration:
                        - Button
                        - Released
                    location:
                      file_id: 0
                      range:
                        start: 636
                        end: 651
                - ~
                - []
                - FollowedBy:
                    constant:
                      Boolean: false
                    expression:
                      SignalCall:
                        signal:
                          id: user_reset_request
                          scope: Output
                        typing: Boolean
                        location:
                          file_id: 0
                          range:
                            start: 665
                            end: 683
                        dependencies:
                          - - user_reset_request
                            - 0
                    typing: Boolean
                    location:
                      file_id: 0
                      range:
                        start: 655
                        end: 683
                    dependencies:
                      - - user_reset_request
                        - 1
              - - Constant:
                    constant:
                      Enumeration:
                        - Button
                        - Pressed
                    location:
                      file_id: 0
                      range:
                        start: 693
                        end: 707
                - ~
                - []
                - MapApplication:
                    function_expression:
                      Call:
                        id: " >= "
                        typing:
                          Abstract:
                            - - Integer
                              - Integer
                            - Boolean
                        location:
                          file_id: 0
                          range:
                            start: 719
                            end: 721
                    inputs:
                      - SignalCall:
                          signal:
                            id: counter
                            scope: Local
                          typing: Integer
                          location:
                            file_id: 0
                            range:
                              start: 711
                              end: 718
                          dependencies:
                            - - counter
                              - 0
                      - MapApplication:
                          function_expression:
                            Call:
                              id: " + "
                              typing:
                                Abstract:
                                  - - Integer
                                    - Integer
                                  - Integer
                              location:
                                file_id: 0
                                range:
                                  start: 736
                                  end: 737
                          inputs:
                            - SignalCall:
                                signal:
                                  id: reset_limit_1
                                  scope: Input
                                typing: Integer
                                location:
                                  file_id: 0
                                  range:
                                    start: 722
                                    end: 735
                                dependencies:
                                  - - reset_limit_1
                                    - 0
                            - SignalCall:
                                signal:
                                  id: reset_limit_2
                                  scope: Input
                                typing: Integer
                                location:
                                  file_id: 0
                                  range:
                                    start: 738
                                    end: 751
                                dependencies:
                                  - - reset_limit_2
                                    - 0
                          typing: Integer
                          location:
                            file_id: 0
                            range:
                              start: 722
                              end: 751
                          dependencies:
                            - - reset_limit_1
                              - 0
                            - - reset_limit_2
                              - 0
                    typing: Boolean
                    location:
                      file_id: 0
                      range:
                        start: 711
                        end: 751
                    dependencies:
                      - - counter
                        - 0
                      - - reset_limit_1
                        - 0
                      - - reset_limit_2
                        - 0
            typing: Boolean
            location:
              file_id: 0
              range:
                start: 605
                end: 758
            dependencies:
              - - user_reset_request
                - 1
              - - counter
                - 0
              - - reset_limit_1
                - 0
              - - reset_limit_2
                - 0
              - - button_state
                - 0
        location:
          file_id: 0
          range:
            start: 574
            end: 759
      user_reset_request_state:
        scope: Output
        id: user_reset_request_state
        signal_type:
          Enumeration: ResetState
        expression:
          Match:
            expression:
              SignalCall:
                signal:
                  id: button_state
                  scope: Input
                typing:
                  Enumeration: Button
                location:
                  file_id: 0
                  range:
                    start: 338
                    end: 350
                dependencies:
                  - - button_state
                    - 0
            arms:
              - - Constant:
                    constant:
                      Enumeration:
                        - Button
                        - Released
                    location:
                      file_id: 0
                      range:
                        start: 362
                        end: 377
                - ~
                - []
                - Constant:
                    constant:
                      Enumeration:
                        - ResetState
                        - Confirmed
                    typing:
                      Enumeration: ResetState
                    location:
                      file_id: 0
                      range:
                        start: 381
                        end: 401
                    dependencies: []
              - - Constant:
                    constant:
                      Enumeration:
                        - Button
                        - Pressed
                    location:
                      file_id: 0
                      range:
                        start: 411
                        end: 425
                - ~
                - []
                - MapApplication:
                    function_expression:
                      Call:
                        id: if_then_else
                        typing:
                          Abstract:
                            - - Boolean
                              - Enumeration: ResetState
                              - Enumeration: ResetState
                            - Enumeration: ResetState
                        location:
                          file_id: 0
                          range:
                            start: 429
                            end: 561
                    inputs:
                      - MapApplication:
                          function_expression:
                            Call:
                              id: " >= "
                              typing:
                                Abstract:
                                  - - Integer
                                    - Integer
                                  - Boolean
                              location:
                                file_id: 0
                                range:
                                  start: 440
                                  end: 442
                          inputs:
                            - SignalCall:
                                signal:
                                  id: counter
                                  scope: Local
                                typing: Integer
                                location:
                                  file_id: 0
                                  range:
                                    start: 432
                                    end: 439
                                dependencies:
                                  - - counter
                                    - 0
                            - SignalCall:
                                signal:
                                  id: reset_limit_1
                                  scope: Input
                                typing: Integer
                                location:
                                  file_id: 0
                                  range:
                                    start: 443
                                    end: 456
                                dependencies:
                                  - - reset_limit_1
                                    - 0
                          typing: Boolean
                          location:
                            file_id: 0
                            range:
                              start: 432
                              end: 456
                          dependencies:
                            - - counter
                              - 0
                            - - reset_limit_1
                              - 0
                      - Constant:
                          constant:
                            Enumeration:
                              - ResetState
                              - InProgress
                          typing:
                            Enumeration: ResetState
                          location:
                            file_id: 0
                            range:
                              start: 488
                              end: 509
                          dependencies: []
                      - Constant:
                          constant:
                            Enumeration:
                              - ResetState
                              - Confirmed
                          typing:
                            Enumeration: ResetState
                          location:
                            file_id: 0
                            range:
                              start: 541
                              end: 561
                          dependencies: []
                    typing:
                      Enumeration: ResetState
                    location:
                      file_id: 0
                      range:
                        start: 429
                        end: 561
                    dependencies:
                      - - counter
                        - 0
                      - - reset_limit_1
                        - 0
            typing:
              Enumeration: ResetState
            location:
              file_id: 0
              range:
                start: 331
                end: 568
            dependencies:
              - - counter
                - 0
              - - reset_limit_1
                - 0
              - - button_state
                - 0
        location:
          file_id: 0
          range:
            start: 287
            end: 569
    unitary_nodes: {}
    location:
      file_id: 0
      range:
        start: 188
        end: 900
    graph:
      vertices:
        button_state:
          id: button_state
          value: Black
          neighbors: []
        counter:
          id: counter
          value: Black
          neighbors:
            - id: res
              weight: 0
            - id: res
              weight: 1
            - id: period
              weight: 0
        period:
          id: period
          value: Black
          neighbors: []
        res:
          id: res
          value: Black
          neighbors:
            - id: button_state
              weight: 0
            - id: button_state
              weight: 1
        reset_limit_1:
          id: reset_limit_1
          value: Black
          neighbors: []
        reset_limit_2:
          id: reset_limit_2
          value: Black
          neighbors: []
        user_reset_request:
          id: user_reset_request
          value: Black
          neighbors:
            - id: user_reset_request
              weight: 1
            - id: counter
              weight: 0
            - id: reset_limit_1
              weight: 0
            - id: reset_limit_2
              weight: 0
            - id: button_state
              weight: 0
        user_reset_request_state:
          id: user_reset_request_state
          value: Black
          neighbors:
            - id: counter
              weight: 0
            - id: reset_limit_1
              weight: 0
            - id: button_state
              weight: 0
component: ~
location:
  file_id: 0
  range:
    start: 0
    end: 900

